### PR TITLE
TFRecords: process and parse with multiple tag databases

### PR DESCRIPTION
**Pull request by pyg4thr1x**
This PR affects preprocessing.py and projectname_input.py

Changelog: 
1. Multiple databases can now be passed at the same time when creating the TFRecords. In that case, the feature dictionary has the following structure: `{'audio', 'tid', 'tags_0', 'tags_1', 'tags_2'}`.

### CODE REVIEW

**Comment by pyg4thr1x** [comment]
I know I should probably finish one PR before working on another one... @hcw1026, I'll submit the fixes you suggested on 'lastfm on steroids' asap

**Comment by pyg4thr1x** [comment]
Changelog:
2. Make use of tf.nn.batch_normalization instead of `f.divide(tf.subtract(features_dict['audio'], mean), tf.sqrt(variance+tf.constant(0.000001)))`

The result is the same.

**Comment by pyg4thr1x** [comment]
Question: why are we not normalising the _batches_ of log mel spectrograms? What changes if instead of         

`# normalize data`
`if audio_format == 'log-mel-spectrogram':`
`.... dataset = dataset.map(_spect_normalization, num_parallel_calls=tf.data.experimental.AUTOTUNE)`
`else:`
`.... dataset = dataset.map(_batch_normalization, num_parallel_calls=tf.data.experimental.AUTOTUNE)`

we do

`# normalize data`
`if audio_format == 'log-mel-spectrogram':`
`.... dataset = dataset.map(_spect_normalization, num_parallel_calls=tf.data.experimental.AUTOTUNE)`
`dataset = dataset.map(_batch_normalization, num_parallel_calls=tf.data.experimental.AUTOTUNE)`

**Comment by hcw1026** [comment]
> I know I should probably finish one PR before working on another one... @hcw1026, I'll submit the fixes you suggested on 'lastfm on steroids' asap

Looks fine if you have tested it.

**Comment by hcw1026** [comment]
> Question: why are we not normalising the batches of log mel spectrograms? What changes if instead of
> if audio_format == 'log-mel-spectrogram':
> .... dataset = dataset.map(_spect_normalization, num_parallel_calls=tf.data.experimental.AUTOTUNE)
> else:
> .... dataset = dataset.map(_batch_normalization, num_parallel_calls=tf.data.experimental.AUTOTUNE)
> we do
> if audio_format == 'log-mel-spectrogram':
> .... dataset = dataset.map(_spect_normalization, num_parallel_calls=tf.data.experimental.AUTOTUNE)
> dataset = dataset.map(_batch_normalization, num_parallel_calls=tf.data.experimental.AUTOTUNE)

This was changed by Calle, since in the paper they normalise log-mel individually

**Comment by hcw1026** [comment]
> Changelog:
> 2. Make use of tf.nn.batch_normalization instead of f.divide(tf.subtract(features_dict['audio'], mean), tf.sqrt(variance+tf.constant(0.000001)))
> The result is the same.

It's keepdims on my laptop...

**Comment by pyg4thr1x** [comment]
> It's keepdims on my laptop...

See the last fix typo commit. I did put back keepdims

**Comment by pyg4thr1x** [comment]
> This was changed by Calle, since in the paper they normalise log-mel individually

I'll test it... It looks sensible to still normalise the whole batch as well, but I woudn't worry about it here

**Comment by pyg4thr1x** [comment]
> Looks fine if you have tested it.

Yes, I have. I am currently generating TFRecords with this branch

**Review by hcw1026 — APPROVED** [review]
